rfd/0089-merge-webapps.md

```diff
@@ -0,0 +1,192 @@
+---
+authors: Jeff Pihach (jeff.pihach@goteleport.com)
+state: draft
+---
+
+# RFD 89 - Merge Webapps into Teleport
+
+## What
+
+- Merge the [webapps](https://github.com/gravitational/webapps) repo into the
+  [teleport](https://github.com/gravitational/teleport) repo.
+- Merge the [webapps.e](https://github.com/gravitational/webapps.e) repo into
+  the [teleport.e](https://github.com/gravitational/teleport.e) repo.
+
+## Why
+
+The version of the web UI and the Teleport are tightly coupled but the
+repositories were created separate to support both Gravity and Teleport. This
+introduced complexities when integrating the two projects.
+
+We have to separately clone and build the web UI and commit the webasset
+changes into the teleport repo whenever we want to represent an update. We also
+have to maintain multiple build systems and ops procedures for the web UI.
+
+Now that Gravity is no longer a product we develop, moving the web UI
+repositories into their respective Teleport repo will save us time and
+confusion.
+
+## Success Criteria
+
+Both [Teleport](https://github.com/gravitational/teleport) and
+[Teleport.e](https://github.com/gravitational/teleport.e) repos contain their
+respective UI counterparts and the build systems correctly build all projects.
+
+## Developer Experience
+
+### Day to day
+
+The Day to day development experience shouldn't change. The webapps project
+will continue to be self-contained and only require access to the nodejs and
+yarn binaries. The locations of the enterprise assets will have to be changed
+but that should remain transparent to the day to day activities.
+
+To work on the enterprise version of the application the developer will have to
+check out the enterprise version of Teleport which will bring with it the
+appropriate webapp assets.
+
+Merging these repositories may cause some disruption when multiple developers
+are working on the same feature across the back and frontend. Using a
+[git worktree](https://git-scm.com/docs/git-worktree) to create a worktree of
+the webapps folder is an elegant way of approaching this issue
+
+### Backports
+
+To avoid having different proceses and build systems depending on the version
+of teleport, all supported release branches (v9, v10, v11) will have their
+respective versions of the webapps repository.
+
+### Build process
+
+The Teleport build process will need to be expanded to include nodejs and yarn
+in order to be able to build the webapp assets and include into the final
+teleport binaries. We will no longer be committing a compiled version of the
+webassets into the teleport repository, the UI will need to be compiled from
+source on the initial build of teleport. A make target will be provided to
+perform a dockerized build of the webassets for those who don't wish to install
+node and yarn on their workstation.
```

Comment:
> Does anyone actually develop the teleport binary without embedding webassets?

I run debug sessions with `DEBUG=1 DEBUG_ASSETS_PATH=/path/to/webassets`. Super helpful to figure out flows that you are not familiar with.

Reply:
> I run debug sessions with `DEBUG=1 DEBUG_ASSETS_PATH=/path/to/webassets`. Super helpful to figure out flows that you are not familiar with.

Could you give an example of a scenario where you'd use this? It's not clear to me why I'd want to use those two env vars over recompiling teleport with embedded webassets or proxy updated Web UI to an existing cluster with `yarn start-teleport --target=https://example.com:3080/web`.

I suppose `DEBUG=1 DEBUG_ASSETS_PATH=…` replaces `yarn start-teleport` if you don't have Node.js tooling installed? Because then you can go to webapps, execute `make build-teleport` and then point `DEBUG_ASSETS_PATH` at assets that were just generated for you through Docker.

Reply:
That is not for Web UI development, but part of the invocation I use to debug the `teleport` binary itself for backend dev. I generally don't care about the webassets version, I just need the server to start.

Reply:
I could potentially see a env variable `NO_UI=1 make full` or something to skip making the UI but I think that's out of scope for now.

[this comment](https://github.com/gravitational/teleport/pull/16413#discussion_r1001123524) outlines how this will work a bit. So one of the steps in the build process will create the built webassets and then the rest of the build will carry-on as it does now. I feel like this will be the quickest way to get this integrated and then we can adjust the process afterwards if we feel it needs adjusting.

Reply:
> That is not for Web UI development, but part of the invocation I use to debug the `teleport` binary itself for backend dev. I generally don't care about the webassets version, I just need the server to start.

@codingllama Ah, so you compile `teleport` without webassets and then use DEBUG & DEBUG_ASSETS_PATH just so that you can start the server?

> (…) one of the steps in the build process will create the built webassets and then the rest of the build will carry-on as it does now. I feel like this will be the quickest way to get this integrated and then we can adjust the process afterwards if we feel it needs adjusting.

@hatched Yeah, it'll definitely be the quickest way. And if other people use a similar workflow for backend dev to what Alan described then it might not be as disruptive.